apps/omg_eth/lib/omg_eth/ethereum_height_monitor.ex

```diff
@@ -0,0 +1,150 @@
+# Copyright 2019-2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.Eth.EthereumHeightMonitor do
+  @moduledoc """
+  Monitors the Ethereum height and raises an alarm when new root chain blocks
+  are not received in a timely manner.
+
+  Note that this module monitors valid block numbers only. Errors are ignored.
+  To monitor connectivity errors, consider using `OMG.Eth.EthereumClientMonitor`.
+  """
+  use GenServer
+  require Logger
+
+  alias OMG.Eth.EthereumClientMonitor
+  alias OMG.Eth.EthereumHeight
+
+  @height_monitor __MODULE__
+
+  @default_interval Application.get_env(:omg_eth, :ethereum_stalled_sync_check_interval_ms)
+
+  @type t :: %__MODULE__{
+          interval: pos_integer(),
+          tref: reference() | nil,
+          alarm_module: module(),
+          raised: boolean(),
+          last_checked_height: non_neg_integer()
+        }
+  defstruct interval: @default_interval,
+            tref: nil,
+            alarm_module: nil,
+            raised: false,
+            last_checked_height: 0
+
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  def init(opts) do
+    _ = Logger.info("Starting Ethereum height monitor.")
+    _ = install_alarm_handler()
+
+    state = %__MODULE__{
+      alarm_module: Keyword.fetch!(opts, :alarm_module)
+    }
+
+    {:ok, state, {:continue, :start_check}}
+  end
+
+  def handle_continue(:start_check, state) do
+    ethereum_height = EthereumHeight.get()
+    stalled = stalled?(ethereum_height, state.last_checked_height)
+    _ = raise_or_clear(state.alarm_module, state.raised, stalled)
+
+    {:ok, tref} = :timer.send_after(state.interval, :check_height)
+    {:ok, %{state | tref: tref, last_checked_height: ethereum_height}}
+  end
+
+  def handle_info(:check_height, state) do
+    ethereum_height = EthereumHeight.get()
+    stalled = stalled?(ethereum_height, state.last_checked_height)
+    _ = raise_or_clear(state.alarm_module, state.raised, stalled)
+
+    {:ok, tref} = :timer.send_after(state.interval, :check_height)
+    {:noreply, %{state | tref: tref, last_checked_height: ethereum_height}}
+  end
+
+  # Handles alarm events from AlarmHandler
+  def handle_cast(:clear_alarm, state) do
+    {:noreply, %{state | raised: false}}
+  end
+
+  def handle_cast(:set_alarm, state) do
+    {:noreply, %{state | raised: true}}
+  end
+
+  defmodule AlarmHandler do
```

Comment:
what's the rationale keeping a module inside a module?

Reply:
I wanted to make clear separation between the monitor and the alarm handler (was a bit hard to distinguish all the callbacks). And the handler is not used outside EthereumHeightMonitor so I placed as a submodule.

Maybe other suggestions?